accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,215 @@
+# ***Support for NuGet authentication plugins deployed via .NET tools***
+
+- Author Name: https://github.com/kartheekp-ms
+- GitHub Issue: https://github.com/NuGet/Home/issues/12567
+
+## Summary
+
+Currently, NuGet utilizes a [cross-platform plugin model](https://learn.microsoft.com/nuget/reference/extensibility/nuget-cross-platform-plugins#supported-operations) which is primarily used for [authentication against private feeds](https://learn.microsoft.com/en-us/nuget/reference/extensibility/nuget-cross-platform-authentication-plugin). It also supports the [package download](https://github.com/NuGet/Home/wiki/NuGet-Package-Download-Plugin) operation.
+
+To accommodate all scenarios involving NuGet client tools, plugin authors will need to create plugins for both `.NET Framework` and `.NET Core`. The following details the combinations of client and framework for these plugins.
+
+| Client tool | Framework |
+|-------------|-----------|
+| Visual Studio | .NET Framework |
+| dotnet.exe | .NET Core |
```

Comment:
I recall that one of the main challenges for making a `dotnet tool` installation work in the `.NET Core` scenario is that NuGet is only looking for a `.dll` file, but the `dotnet tool` installation process emits a `.exe` wrapper into the default tool directory.

Reply:
> but the dotnet tool installation process emits a .exe wrapper into the default tool directory.

Only on Windows. On Mac and Linux it's an executable without an extension.

Reply:
either way, the issue is that NuGet requires that the plugin be a .dll file invokable as `dotnet <whatever>.dll`, rather than allowing an arbitrary executable

Reply:
On Unixlikes, "an arbitrary executable" should be able to include an arbitrary shell script with a `#!` header (implementing the relatively complex protocol may be difficult in a shell script, but very doable if the named interpreter is, say, Python)

Reply:
> the issue is that NuGet requires that the plugin be a .dll file invokable as dotnet <whatever>.dll

That is the current situation. However, if the credential provider is installed as a .NET Tool (which is what this spec is all about), NuGet will invoke `<whatever>.exe` on Windows, regardless of whether the app is targeting `.NET Framework` or `.NET Core`. On Linux and Mac, NuGet will invoke `<whatever>` assuming the file has the executable bit set. Does that answer your question, @jmyersmsft?

Reply:
> On Unixlikes, "an arbitrary executable" should be able to include an arbitrary shell script with a #! header (implementing the relatively complex protocol may be difficult in a shell script, but very doable if the named interpreter is, say, Python)

Can you please explain this comment so that I can understand better?

Reply:
on Unixlikes, shell scripts can be run in either of two ways:

1. Run the *interpreter*, passing the name of the script as an argument, for example:

   ```shell
   $ /bin/bash -v my/script.sh --some-script-argument
   ```

2. If the script has the execute bit set and begins with `#!name-of-interpreter --some-interpreter-argument`, it can be invoked directly, for example, if you have the script `my/script.sh`:
   ```bash
   #!/bin/bash -v
   echo "Hello, world!"
   ```
   and you run the command:
   ```shell
   $ my/script.sh --some-script-argument
   ```
   the system will *actually* run `/bin/bash -v my/script.sh --some-script-argument`. This is handled at the system level. Running the script as a program in the "usual way" will automatically do the right thing. The file extension  is irrelevant and many shell scripts intended to run this way have no extension (just as with proper executables)

In the context of this feature, this means that it would be nice if NuGet didn't try to do things like detect PE- or ELF- format files. The plugin protocol is probably a little more complicated than you'd want to implement in bash (or worse, POSIX sh), but would be quite plausible in something like Python.

Reply:
Thanks, for the additional information.

> In the context of this feature, this means that it would be nice if NuGet didn't try to do things like detect PE- or ELF- format files.

In the current proposal, the plan is to invoke the credential provider installed as a .NET tool, following the guidance provided in the [.NET SDK documentation](https://learn.microsoft.com/dotnet/core/tools/global-tools-how-to-use#use-the-tool-as-a-global-tool-installed-in-a-custom-location).

   On Windows:

   ```console
   c:\dotnet-tools\botsay hello from the bot
   ```

   On Linux or macOS:

   ```console
   ~/bin/botsay hello from the bot
  ```

NuGet currently invokes plugins in a [separate process](https://github.com/NuGet/NuGet.Client/blob/0d9fe5cc3d353321d323d110fdcc4aa383d7c1ce/src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs#L155-L181). The plan is to continue doing so, with the exception of plugins installed as .NET Tools. For these plugins, on `Windows` the .exe file will be executed in a separate process, regardless of whether it’s using .NET Framework or .NET Core. On `Linux and Mac`, the file with the executable bit set will be launched in a separate process, along with the required parameters.